### DISCUSSION TITLE
api-vip / k3s_registration_address

### DISCUSSION THREAD

**Discussion by tuxpeople** [PyratLabs/ansible-role-k3s#154] [Q&A] [answered]
Hi

This topic is about an mutli-master setup with internal etcd.

In my current setup, I use Ansible to to deploy a pair of HAproxy nodes and keepalived for them, and use the keepalived-VIP later for this role as `k3s_registration_address` and also as a part of `tls-san`.

I'd like to get rid of them. I already stopped using them for ingress, as ingress is now Traefik with MetalLB. I luse flux to deplay the rest of my cluster, and I also would like to deploy `keepalived-ingress-vip` (https://github.com/janeczku/keepalived-ingress-vip) or kube-vip there. But I need that already earlier, for `k3s_registration_address` right? 

Can I use `k3s_server_manifests_templates` or `k3s_server_manifests_urls` for this? Does it get deployed before the additional nodes join? Or is `k3s_registration_address` only for the initial join and it will never be used again? If so, I could just use first node's IP for joining and deploy vip later.

Br
Thomas

**Comment by xanmanning** [comment, marked as answer]
`k3s_registration_address` is the address that worker nodes use to connect to the control plane on initial join of the cluster, if I recall though, if you restart a worker and it isn't able to connect to it then it does fail to start.

Or at least it did at some point in the past, who knows if it still does?

**Reply by tuxpeople** [reply]
Is `k3s_server_manifests_urls` being deployed before other masters joining the cluster? If so, I could use it for `kube-vip` or similiar and use that IP for `k3s_registration_address`. Does the role wait a while for `k3s_registration_address` to become available?

**Reply by xanmanning** [reply]
Yep, `k3s_server_manifests_urls` is deployed before other "masters" join - they are actually installed before K3s is even installed: https://github.com/PyratLabs/ansible-role-k3s/blob/0bdf5383efb34d417082fb3f2d0b49c19130d27e/tasks/state-installed.yml#L37-L40

The role doesn't explicitly wait for `k3s_registration_address`, it waits for the initial "master" to start and be listening on the right port:
https://github.com/PyratLabs/ansible-role-k3s/blob/main/tasks/validate/state/control-plane.yml

However post install it waits for all the workers to be ready:

https://github.com/PyratLabs/ansible-role-k3s/blob/0bdf5383efb34d417082fb3f2d0b49c19130d27e/tasks/validate/state/nodes.yml#L3-L20
